### ISSUE TITLE
[dev.icinga.com #173] fix duplicate entries in table servicechecks (mysql)

### ISSUE THREAD

**Issue by icinga-migration** [Icinga/icinga-core#82]
This issue has been migrated from Redmine: https://dev.icinga.com/issues/173

**Created by mfriedrich on 2009-11-03 11:37:31 +00:00**

Assignee: _mfriedrich_
Status: _Resolved_  (closed on _2009-11-03 12:01:07 +00:00_)
Target Version: _1.0 Stable_
Last Update: _2009-11-03 12:01:07 +00:00 (in Redmine)_

---

my comment from nagios-devel
-------------------------------

Hi,

Øyvind Nordang wrote:
> 
> When I check the servicechecks table each check of a service is listed
> twice, but not with the same information in both rows.
> 
I've noticed that myself and that seems to be a bug. Maybe it concerns the unique constraint where the UPDATE should happen. But it could also be that some data is sent incorrectly and therefore no matching can be done.
> In dbhandlers.c on line 1518 all I can think of is that ON DUPLICATE
> KEY UPDATE does not work as it should. I haven't had time to look at
> the code and programming isn't my strong side.
> 
> if(asprintf(&buf,"INSERT INTO %s SET %s, command\_object\_id='%lu',
> command\_args='%s', command\_line='%s' ON DUPLICATE KEY UPDATE %s"
> ,ndo2db\_db\_tablenames\[NDO2DB\_DBTABLE\_SERVICECHECKS\]
> ,buf1
> ,command\_id
> ,es\[0\]
> ,es\[1\]
> ,buf1
> )==-1)
> buf=NULL;
> 
Also the creation of "buf1" is interesting. The approach of building the query buffer for both INSERT and UPDATE might lead into a real error too. The only difference is that the INSERT statement gets command\_object\_id, command\_args and command\_line

The defined keys in db/mysql.sql

KEY \`instance\_id\` (\`instance\_id\`),
KEY \`service\_object\_id\` (\`service\_object\_id\`),
KEY \`start\_time\` (\`start\_time\`)

Kind of strange, no unique definition ... but let's step further into analysis.

Problem regarding that routine is, that the constraint never matches.
I'm trying to resolve this error from Icinga IDOUtils where the error exists too. The main differences are that the queries are normalized (INSERT INTO .. () VALUES () instead of INSERT INTO .. SET ..=..) but the constraints are the same as in NDOUtils. Also the debugoutput is quite enhanced to see what's going on.

I've run several tests with Postgres and Oracle, and regarding especially right now Oracle, the defined constraint also matches and entries do not get duplicated. Postgres also works fine.

The reason why this is happening is quite simple - the queries point exactly to the unique constraint.

Postgres:
UPDATE %s SET ... WHERE instance\_id=%lu AND service\_object\_id=%lu AND start\_time=%s AND start\_time\_usec=%lu

Oracle:
MERGE INTO %s USING DUAL ON (instance\_id=%lu AND service\_object\_id=%lu AND start\_time=%s AND start\_time\_usec=%lu) WHEN MATCHED THEN UPDATE SET ... WHEN NOT MATCHED THEN INSERT ... VALUES ...

Conclusion to that is that it is a MySQL problem regarding the defined keys and the ON DUPLICATE KEY statement.

To step further into that, I've only selected the columns which are necessary to see the difference::

mysql&gt; select servicecheck\_id, service\_object\_id, start\_time, start\_time\_usec, end\_time, end\_time\_usec, command\_object\_id, execution\_time from icinga\_servicechecks;

*-----------------*------------------~~*--------------------~~*----------------~~*---------------------*--------------~~*------------------~~*---------------~~+
| servicecheck\_id | service\_object\_id | start\_time | start\_time\_usec | end\_time | end\_time\_usec | command\_object\_id | execution\_time |

*-----------------*------------------~~*--------------------~~*----------------~~*---------------------*--------------~~*------------------~~*---------------~~+

This block should be have been updated e.g.

|     |      |                     |        |                     |     |      |     |
|-----|------|---------------------|--------|---------------------|-----|------|-----|
| 521 | 6421 | 2009-11-03 10:34:58 | 245848 | 1970-01-01 01:00:00 | 0   | 6193 | 0   |
| 522 | 6422 | 2009-11-03 10:34:58 | 339129 | 1970-01-01 01:00:00 | 0   | 6193 | 0   |
| 523 | 6423 | 2009-11-03 10:35:02 | 188397 | 1970-01-01 01:00:00 | 0   | 6193 | 0   |
| 524 | 6424 | 2009-11-03 10:35:02 | 329014 | 1970-01-01 01:00:00 | 0   | 6193 | 0   |
| 525 | 6425 | 2009-11-03 10:35:03 | 91160  | 1970-01-01 01:00:00 | 0   | 6193 | 0   |
| 526 | 6307 | 2009-11-03 10:35:03 | 305725 | 1970-01-01 01:00:00 | 0   | 6193 | 0   |
| 527 | 6426 | 2009-11-03 10:35:04 | 57256  | 1970-01-01 01:00:00 | 0   | 6193 | 0   |
| 528 | 6427 | 2009-11-03 10:35:04 | 150767 | 1970-01-01 01:00:00 | 0   | 6193 | 0   |
| 529 | 6429 | 2009-11-03 10:35:05 | 172237 | 1970-01-01 01:00:00 | 0   | 6193 | 0   |
| 530 | 6428 | 2009-11-03 10:35:05 | 241257 | 1970-01-01 01:00:00 | 0   | 6193 | 0   |

But it doesn't, you can see the INSERT statements right after a while in the history.

|     |      |                     |        |                     |        |      |         |
|-----|------|---------------------|--------|---------------------|--------|------|---------|
| 531 | 6421 | 2009-11-03 10:34:58 | 245848 | 2009-11-03 10:35:02 | 798210 | 0    | 4.55236 |
| 532 | 6422 | 2009-11-03 10:34:58 | 339129 | 2009-11-03 10:35:02 | 801988 | 0    | 4.46286 |
| 533 | 6430 | 2009-11-03 10:35:16 | 721993 | 1970-01-01 01:00:00 | 0      | 6193 | 0       |
| 534 | 6423 | 2009-11-03 10:35:02 | 188397 | 2009-11-03 10:35:06 | 395320 | 0    | 4.20692 |
| 535 | 6424 | 2009-11-03 10:35:02 | 329014 | 2009-11-03 10:35:06 | 786733 | 0    | 4.45772 |
| 536 | 6307 | 2009-11-03 10:35:03 | 305725 | 2009-11-03 10:35:08 | 45368  | 0    | 4.73964 |
| 537 | 6425 | 2009-11-03 10:35:03 | 91160  | 2009-11-03 10:35:08 | 389047 | 0    | 5.29789 |
| 538 | 6426 | 2009-11-03 10:35:04 | 57256  | 2009-11-03 10:35:08 | 674087 | 0    | 4.61683 |
| 539 | 6427 | 2009-11-03 10:35:04 | 150767 | 2009-11-03 10:35:09 | 322330 | 0    | 5.17156 |
| 540 | 6429 | 2009-11-03 10:35:05 | 172237 | 2009-11-03 10:35:09 | 524681 | 0    | 4.35244 |
| 541 | 6428 | 2009-11-03 10:35:05 | 241257 | 2009-11-03 10:35:09 | 900865 | 0    | 4.65961 |
| 542 | 6431 | 2009-11-03 10:35:17 | 26969  | 1970-01-01 01:00:00 | 0      | 6193 | 0       |
| 543 | 6432 | 2009-11-03 10:35:17 | 133886 | 1970-01-01 01:00:00 | 0      | 6193 | 0       |
| 544 | 6430 | 2009-11-03 10:35:16 | 721993 | 2009-11-03 10:35:20 | 996513 | 0    | 4.27452 |
| 545 | 6431 | 2009-11-03 10:35:17 | 26969  | 2009-11-03 10:35:21 | 391540 | 0    | 4.36457 |
| 546 | 6432 | 2009-11-03 10:35:17 | 133886 | 2009-11-03 10:35:21 | 396431 | 0    | 4.26255 |

Let's focus on service\_object\_id=6421 (i've cleared the IP address)

The first query generates the entry where end\_time is 0 ... FROM\_UNIXTIME(0), 0 ... start\_time is 1257240898 (= 2009-11-03 10:34:58 GMT+1)

|     |      |                     |        |                     |     |      |     |
|-----|------|---------------------|--------|---------------------|-----|------|-----|
| 521 | 6421 | 2009-11-03 10:34:58 | 245848 | 1970-01-01 01:00:00 | 0   | 6193 | 0   |

\[1257240916.978113\] \[002.0\] \[pid=2673\] INSERT INTO icinga\_servicechecks (instance\_id, service\_object\_id, check\_type, current\_check\_attempt, max\_check\_attempts, state, state\_type, start\_time, start\_time\_usec, end\_time, end\_time\_usec, timeout, early\_timeout, execution\_time, latency, return\_code, output, long\_output, perfdata, command\_object\_id, command\_args, command\_line) VALUES (1, 6421, 0, 1, 3, 0, 1, FROM\_UNIXTIME(1257240898), 245848, FROM\_UNIXTIME(0), 0, 60, 0, 0.000000, 0.245000, 0, 'PING OK - Packet loss = 0, RTA = 17\\.43 ms', '', '', 6193, '100\\.0,20!500\\.0,60', '/opt/icinga/libexec/check\_ping -H 1\\.2\\.3\\.4 -w 100\\.0,20 -c 500\\.0,60% -p 5 ~~4') ON DUPLICATE KEY UPDATE check\_type='0', current\_check\_attempt='1', max\_check\_attempts='3', state='0', state\_type='1', start\_time=FROM\_UNIXTIME(1257240898), start\_time\_usec='245848', end\_time=FROM\_UNIXTIME(0), end\_time\_usec='0', timeout='60', early\_timeout='0', execution\_time='0.000000', latency='0.245000', return\_code='0', output='PING OK~~ Packet loss = 0%, RTA = 17\\.43 ms', long\_output='', perfdata=''

For that check, we are waiting for end\_time and end\_time\_usec and truly need the UPDATE condition, the INSERT will create this faulty and duplicate entry (short version!).

|     |      |                     |        |                     |        |     |         |
|-----|------|---------------------|--------|---------------------|--------|-----|---------|
| 531 | 6421 | 2009-11-03 10:34:58 | 245848 | 2009-11-03 10:35:02 | 798210 | 0   | 4.55236 |

\[1257240917.081916\] \[002.0\] \[pid=2673\] INSERT INTO icinga\_servicechecks (instance\_id, service\_object\_id, check\_type, current\_check\_attempt, max\_check\_attempts, state, state\_type, start\_time, start\_time\_usec, end\_time, end\_time\_usec, timeout, early\_timeout, execution\_time, latency, return\_code, output, long\_output, perfdata, command\_object\_id, command\_args, command\_line) VALUES (1, 6421, 0, 1, 3, 0, 1, FROM\_UNIXTIME(1257240898), 245848, FROM\_UNIXTIME(1257240902), 798210, 60, 0, 4.552360, 0.245000, 0, 'PING OK - Packet loss = 0, RTA = 75\\.77 ms', '', '', 0, '', '') ON DUPLICATE KEY UPDATE check\_type='0', current\_check\_attempt='1', max\_check\_attempts='3', state='0', state\_type='1', start\_time=FROM\_UNIXTIME(1257240898), start\_time\_usec='245848', end\_time=FROM\_UNIXTIME(1257240902), end\_time\_usec='798210', timeout='60', early\_timeout='0', execution\_time='4.552360', latency='0.245000', return\_code='0', output='PING OK - Packet loss = 0, RTA = 75\\.77 ms', long\_output='', perfdata=''

As you may see on a full output on that row, the INSERT puts empty command\_line, empty command\_object\_id etc into the DB which is wrong.

Now let's compare on the unique constraint, where ON DUPLICATE KEY shoud trigger on the UPDATE statement:

instance\_id - same!
service\_object\_id - same!
start\_time - same!

Well regarding the fact that the unique constraint is missing and instead normal keys are defined, let's turn the other way around - why are the hostchecks running fine without duplicate entries? Hmmm, because the unique key is defined in a correct manner.

servicechecks:
KEY \`instance\_id\` (\`instance\_id\`),
KEY \`service\_object\_id\` (\`service\_object\_id\`),
KEY \`start\_time\` (\`start\_time\`)

hostchecks:
UNIQUE KEY \`instance\_id\` (\`instance\_id\`,\`host\_object\_id\`,\`start\_time\`,\`start\_time\_usec\`)

Whoever did that to the SQL-Statement during table creation......

Correct way of getting that to work will be defining a real unique key and then ON DUPLICATE KEY will trigger the UPDATE statement.

I will work on a patch and provide it when finished testing.

Kind regards,
Michael
